tests/turnstile_complex.c

```diff
@@ -0,0 +1,171 @@
+  do {                                                                         \
+    WITH_SPINLOCK(td->td_spin) {                                               \
+      sched_lend_prio(td, prio);                                               \
+    }                                                                          \
+  } while (0)
+
+#define unlend_prio(td, prio)                                                  \
+  do {                                                                         \
+    WITH_SPINLOCK(td->td_spin) {                                               \
+      sched_unlend_prio(td, prio);                                             \
+    }                                                                          \
+  } while (0)
+
+#define set_base_prio(td, prio)                                                \
+  do {                                                                         \
+    WITH_SPINLOCK(td->td_spin) {                                               \
+      td->td_base_prio = prio;                                                 \
+    }                                                                          \
+  } while (0)
+
+static mtx_t *mtx1 = &MTX_INITIALIZER(MTX_DEF);
+static mtx_t *mtx2 = &MTX_INITIALIZER(MTX_DEF);
+static thread_t *td[T];
+static volatile bool low_prio_mtx1_acquired;
+static volatile bool med_prio_mtx1_acquired;
+static volatile bool high_prio_mtx2_acquired;
+
+enum {
+  /* Priorities are multiplies of RunQueue_PriorityPerQueue
+   * so that each priority matches different run queue. */
+  LOW = 0,
+  MED = RQ_PPQ,
+  HIGH = 2 * RQ_PPQ
+};
+
+/* td3 */
+static void high_prio_task(void *arg) {
+  WITH_MTX_LOCK (mtx2) {
+    /* When we get here, low_prio_task and med_prio_task1 should've
+       got their mutexes and all priorities should be in initial state. */
+    high_prio_mtx2_acquired = 1;
+    assert(low_prio_mtx1_acquired == 1);
+    assert(med_prio_mtx1_acquired == 1);
+
+    assert_priorities(LOW, MED, MED, HIGH);
+  }
+}
+
+/* td1 */
+static void med_prio_task1(void *arg) {
+  WITH_NO_PREEMPTION {
+    WITH_MTX_LOCK (mtx2) {
+      WITH_MTX_LOCK (mtx1) {
+        /* Td0 (low_prio_task) has released mtx1.
+         * Td3 (high_prio_task) is waiting for mtx2. */
+        med_prio_mtx1_acquired = 1;
+        assert(low_prio_mtx1_acquired == 1);
+        assert(high_prio_mtx2_acquired == 0);
+
+        assert(thread_self()->td_prio == HIGH);
+        assert(thread_self()->td_flags & TDF_BORROWING);
+      }
+    }
+  }
+
+  /* Enabling preemption should've preempted us and switched to
+   * high_prio_task. */
+  assert(high_prio_mtx2_acquired == 1);
+  assert(!(thread_self()->td_flags & TDF_BORROWING));
+  assert_priorities(LOW, MED, MED, HIGH);
+}
+
+/* td2 */
+static void med_prio_task2(void *arg) {
+  /* Without turnstile mechanism this assert would fail. */
+  assert(high_prio_mtx2_acquired);
```

Comment:
Priority propagation could work only partially though - for example only propagate to the thread owning the mutex (and not to the threads blocking that thread). The test would still fail but not necessarily in this way. I think it would randomly fail either here or on the assert at 88: `assert_priorities(HIGH, HIGH, MED, HIGH)` (`td0`, after second `thread_yield`. Look at 78: `assert_priorities(MED, MED, MED, HIGH)`: with single thread propagation this assert could be passed, `td3` would bump `td1` to `HIGH` but `td2` would still be locked and the only runnable threads would be `td0` and `td2`, both with `MED` priority.

Reply:
My comment doesn't say what would happen if partial priority propagation was implemented.
It says that if you have no priority propagation in mutexes, `med_prio_task2` would run before `high_prio_task` acquires `mtx2`.